### PR TITLE
VAULT-17590 Add failure notifications for OSS builds

### PR DESCRIPTION
**Pull request by VioletHynes**
This uses GitHub secrets as Vault is not yet available in the `vault` repo. The `vault-enterprise` file will use the secret in Vault.

Upon a failure on `main` or a release branch, a failure notification will be sent to `#feed-vault-ci-official`. Some examples are in that channel already, although instead of saying e.g. `21561/merge` it will say the branch it was from, e.g. `main`, or `release/1.14.x` etc.

What's left is to merge this PR, do an OSS merge, then update these notifications so that they use the CI Vault for Ent and say they're an ent run etc. Note that while this will mean we do have 'drift', we have differences in these files for OSS/ent already, so this is 'expected'.

### CODE REVIEW

File: vault/request_handling.go
Comment (VioletHynes) [inline vault/request_handling.go:1]: The changes to this file are intentional (to test the notification) and will be removed before this PR is merged.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -259,3 +259,25 @@ jobs:
...
+      contents: read
+    strategy:
+      fail-fast: false
+    needs:
+      - completed-successfully
+    steps:
+      - name: send-notification
+        uses: slackapi/slack-github-action@007b2c3c751a190b6f0f040e47ed024deaa72844 # v1.23.0
+        # We intentionally aren't using the following here since it's from an internal repo
+        # uses: hashicorp/cloud-gha-slack-notifier@730a033037b8e603adf99ebd3085f0fdfe75e2f4 #v1
```
Comment (VioletHynes) [inline .github/workflows/build.yml:277]: The only difference here is that we're providing `SLACK_BOT_TOKEN` as an environment variable, but I thought I'd point it out.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -319,3 +319,26 @@ jobs:
...
+      - name: send-notification
+        uses: slackapi/slack-github-action@007b2c3c751a190b6f0f040e47ed024deaa72844 # v1.23.0
+        # We intentionally aren't using the following here since it's from an internal repo
+        # uses: hashicorp/cloud-gha-slack-notifier@730a033037b8e603adf99ebd3085f0fdfe75e2f4 #v1
+        env:
+          SLACK_BOT_TOKEN: ${{ secrets.SLACK_BOT_TOKEN }}
+        with:
+          channel-id: "C05AABYEA9Y" # sent to #feed-vault-ci-official
+          payload: |
+            {"text":"OSS test failures on ${{ github.ref_name }}","blocks":[{"type":"header","text":{"type":"plain_text","text":":flashing-light: OSS test failures :flashing-light:","emoji":true}},{"type":"divider"},{"type":"section","text":{"type":"mrkdwn","text":"test(s) failed on ${{ github.ref_name }}"},"accessory":{"type":"button","text":{"type":"plain_text","text":"View Failing Workflow","emoji":true},"url":"${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}"}}]}
```
Comment (kubawi) [inline .github/workflows/ci.yml:343]: 🤔 Same as above 😄 

```suggestion
            {"text":"OSS test failures on ${{ github.ref_name }}","blocks":[{"type":"header","text":{"type":"plain_text","text":":rotating_light: OSS test failures :rotating_light:","emoji":true}},{"type":"divider"},{"type":"section","text":{"type":"mrkdwn","text":"test(s) failed on ${{ github.ref_name }}"},"accessory":{"type":"button","text":{"type":"plain_text","text":"View Failing Workflow","emoji":true},"url":"${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}"}}]}
```

**Review by kubawi — APPROVED** [review]
LGTM, but I left a minor comment.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -259,3 +259,25 @@ jobs:
...
+        env:
+          SLACK_BOT_TOKEN: ${{ secrets.SLACK_BOT_TOKEN }}
+        with:
+          channel-id: "C05AABYEA9Y" # sent to #feed-vault-ci-official
+          payload: |
+            {"text":"OSS build failures on ${{ github.ref_name }}","blocks":[{"type":"header","text":{"type":"plain_text","text":":flashing-light: OSS build failures :flashing-light:","emoji":true}},{"type":"divider"},{"type":"section","text":{"type":"mrkdwn","text":"build(s) failed on ${{ github.ref_name }}"},"accessory":{"type":"button","text":{"type":"plain_text","text":"View Failing Workflow","emoji":true},"url":"${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}"}}]}
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply .github/workflows/build.yml:283]: Good feedback! I don't think this is a nitpick at all. I'll update :)

Comment (miagilepner) [inline .github/workflows/build.yml]: when testing this payload on [the slack block kit builder](https://app.slack.com/block-kit-builder/T024UT03C#%7B%22text%22:%22OSS%20build%20failures%20on%20main%22,%22blocks%22:%5B%7B%22type%22:%22header%22,%22text%22:%7B%22type%22:%22plain_text%22,%22text%22:%22:flashing-light:%20OSS%20build%20failures%20:flashing-light:%22,%22emoji%22:true%7D%7D,%7B%22type%22:%22divider%22%7D,%7B%22type%22:%22section%22,%22text%22:%7B%22type%22:%22mrkdwn%22,%22text%22:%22build(s)%20failed%20on%20main%22%7D,%22accessory%22:%7B%22type%22:%22button%22,%22text%22:%7B%22type%22:%22plain_text%22,%22text%22:%22View%20Failing%20Workflow%22,%22emoji%22:true%7D,%22url%22:%22https://github.com/hashicorp/vault/actions/runs/1234%22%7D%7D%5D%7D), I see that it errors with "invalid additional property 'text'". Is that a problem?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -319,3 +319,26 @@ jobs:
     steps:
       - run: |
           tr -d '\n' <<< '${{ toJSON(needs.*.result) }}' | grep -q -v -E '(failure|cancelled)'
+
+  notify-tests-completed-failures:
+    if: ${{ always() && needs.tests-completed.result == 'failure' && (github.head_ref == github.ref_name == 'main' || startsWith(github.ref_name, 'release/')) }}
```
Comment (miagilepner) [inline .github/workflows/ci.yml:324]: `github.head_ref == github.ref_name == 'main'` is this intentional?

Comment (VioletHynes) [reply .github/workflows/ci.yml:324]: Good catch! It's not. For context, `head_ref` gives you the branch name for PRs, whereas `ref_name` gives the branch name for non-PRs. I messed this up when switching between testing and having this ready to go.

There, strangely, isn't a single property that gives you the branch name universally, so I had to do it this way.

File: .github/workflows/build.yml (hunk 3 above)
Comment (VioletHynes) [reply .github/workflows/build.yml]: Huh, weird. `text` doesn't actually change the message, but essentially provides info for screen-readers etc. When I was first adding these notifications, I got a warning if I didn't include it. I think we're good!

Comment (VioletHynes) [reply .github/workflows/build.yml]: I think the Slack GHA must use a different format to the builder you linked

Comment (VioletHynes) [reply .github/workflows/build.yml]: I think maybe this is using 'approach 2' as listed on this README: https://github.com/slackapi/slack-github-action

**Pull request merged**